### PR TITLE
vi: Translate the std library String type

### PR DESCRIPTION
**Pull request by l0ngvh**
I couldn't find an effective way to translate `Trait` and `implement`, so both terms are going to be used as is in the translation.

### CODE REVIEW

File: po/vi.po
Comment (qu-ngx) [inline po/vi.po:6138]: String can be translated to Xâu ký tự. If you are unsure about the term translation, you can translate using the following format "English Term (Bản dịch tiếng Việt)" and proceed to go on to the part that can be easily translated. 

So it should be the following the format:
"[`String (Xâu ký tự)`] là kiểu (https://doc.rust-lang.org/std/string/struct.String.html) là kiểu " 
"dữ liệu chuẩn sử dụng để lưu trữ chuỗi UTF-8 trên bộ nhớ đệm xâu ký tự (buffer string) sử dụng cấu trúc chất đống (heap), có thể được mở rộng:"

File: po/vi.po
Comment (qu-ngx) [inline po/vi.po:6217]: "implement" not translated?

File: po/vi.po
Comment (qu-ngx) [inline po/vi.po:6225]: "trait" not translated?

File: po/vi.po
Comment (qu-ngx) [inline po/vi.po:6248]: remember to translate "String".

File: po/vi.po
Comment (qu-ngx) [inline po/vi.po:6268]: "xâu con" might be better than "chuỗi con" (just a suggestion). Instead of using "nằm trên" try out "nằm ngoài"? Cái chuỗi này nằm ngoài chuỗi kia sounds a bit more logical ?

File: po/vi.po
Comment (qu-ngx) [inline po/vi.po:6281]: "Trait" not translated

File: po/vi.po
Comment (qu-ngx) [inline po/vi.po:6175]: Translate String?

File: po/vi.po
Comment (qu-ngx) [inline po/vi.po:6176]: Translate "String"?

**Review by ghost — CHANGES_REQUESTED** [review]
I have left some comments and feedbacks before you are able to merge. You should review and change your code before it is merged. Tks!

**Comment by vohoanglong0107** [comment]
@qu-ngx I'm quite unsure about translating `implement` and `trait`, as there are no Vietnamese terms that accurately express them. The closest translation I could think of for `implement` is `kế thừa`, but it's more related to the OOP term `inherit` than `implement`.

**Comment by qu-ngx** [comment]
> @qu-ngx I'm quite unsure about translating `implement` and `trait`, as there are no Vietnamese terms that accurately express them. The closest translation I could think of for `implement` is `kế thừa`, but it's more related to the OOP term `inherit` than `implement`.

Implement is also áp dụng, thi hành, trang bị, .... Trait aka đặc điểm, tính trạng, trạng thái. If you are not so sure, I think you can use some English dictionary (ie Meriam Webster) for the synonyms. English has many different shades so it is not tied to only one meaning at all.

**Comment by vohoanglong0107** [comment]
@qu-ngx Also, when translating type name, such as `String`, I believe translate `String` the first time it appearse is good, but it shouldn't be translated in subsequent appearances, as it can have cause confusion. For example, in Rust `&str` and `String` can be translated as `Xâu` in Vietnamese, but using only `Xâu` in the translation will cause confusion, as the students might be unsure which type to use.

**Comment by qu-ngx** [comment]
> @qu-ngx Also, when translating type name, such as `String`, I believe translate `String` the first time it appearse is good, but it shouldn't be translated in subsequent appearances, as it can have cause confusion. For example, in Rust `&str` and `String` can be translated as `Xâu` in Vietnamese, but using only `Xâu` in the translation will cause confusion, as the students might be unsure which type to use.

It still has translation in Vietnamese and also String translation is a well-known Vietnamese word. That is why when I say translate you put in the brackets. You can refer to the other successful PRs of other people as well if you want. It has no cost and no harm in doing it as not everyone learn it in English first time. That is why both is better than just one.

**Comment by qu-ngx** [comment]
> @qu-ngx Also, when translating type name, such as `String`, I believe translate `String` the first time it appearse is good, but it shouldn't be translated in subsequent appearances, as it can have cause confusion. For example, in Rust `&str` and `String` can be translated as `Xâu` in Vietnamese, but using only `Xâu` in the translation will cause confusion, as the students might be unsure which type to use.

Also please remember you are doing the translation for everyone so clarity in the words and meanings have a significant role in why you are doing translations. This feedbacks are not set to target you, but to help you find a better way to make your translation as clear as possible to the other Vietnamese peers.

**Comment by vohoanglong0107** [comment]
> Implement is also áp dụng, thi hành, trang bị, .... Trait aka đặc điểm, tính trạng, trạng thái. If you are not so sure, I think you can use some English dictionary (ie Meriam Webster) for the synonyms. English has many different shades so it is not tied to only one meaning at all.

@qu-ngx I could use these translations, but they wouldn't maintain the original meaning. For example, how would `Deserializer implement Sync Trait` be translated?
IMO, we shouldn't force a translation on a technical terms. [Rust Tiếng Việt](https://rust-tieng-viet.github.io/basic/trait/index.html) agree with me on this point, as they also don't translate `Trait` and `implement`

**Comment by vohoanglong0107** [comment]
> Also, please remember you are doing the translation for everyone so clarity in the words and meanings have a significant role in why you are doing translations. These feedbacks are not set to target you, but to help you find a better way to make your translation as clear as possible to the other Vietnamese peers.

I understand, which is why I want to discuss with you the best way to translate this. I want to introduce this book to my Vietnamese friends, and I don't want to confuse them while reading the book.

**Comment by vohoanglong0107** [comment]
@qu-ngx I have opened a new discussion to talk about this [here](https://github.com/google/comprehensive-rust/discussions/2062)

File: po/vi.po
Comment (qu-ngx) [reply po/vi.po:6217]: this implements is not a built-in in C++ btw. implements (some Vietnamese translation) should be good

**Review by ghost — CHANGES_REQUESTED** [review]
I have removed some comments on translation for built-ins. Can you help me out with the translation? Tkiu so much!

**Comment by vohoanglong0107** [comment]
@qu-ngx I think some previous comments are staled after https://github.com/google/comprehensive-rust/discussions/2062#discussioncomment-9335645. Could you give this another go?

**Comment by qu-ngx** [comment]
> @qu-ngx I think some previous comments are staled after [#2062 (comment)](https://github.com/google/comprehensive-rust/discussions/2062#discussioncomment-9335645). Could you give this another go?

Oh yea, sorry I forgot to remove some on comments. Can you check it for me. And btw the other "implements" is an action word. It isn't the name of the built-in `implement`. Can you help translate that one?

File: po/vi.po
Comment (vohoanglong0107) [reply po/vi.po:6217]: `Deref<Target = T>` is a trait, so this `implement` should be a technical term in Rust

Comment (qu-ngx) [reply po/vi.po:6217]: "When a type implements `Deref<Target = T>`". What I understand here is  I think "implements" here is not a technical term in Rust since it has "s" in the word? Are you sure it's not? This is what I got from using a translator:
"Khi một kiểu biến triển khai `Deref<Target = T>`, trình biên dịch sẽ cho phép bạn .." 

The `implement` you referred to is without s. But in that sentence implement also go with "s". So is it a technical term or not?

**Comment by vohoanglong0107** [comment]
@qu-ngx Are you referring to another `implement`?
Because `Deref<Target = T>` is a trait, you can learn more about this trait [here](https://doc.rust-lang.org/std/ops/trait.Deref.html), `implement` in  `implement Deref<Target = T>` isn't being used as English action verb. What `implement Deref<Target = T>` actually means is, any type that `implement Deref<Target = T>` will have a method called `deref`, which in many cases means that the type allows user to access the underlying type.
I hope this clear things up a bit.

**Review by ghost — COMMENTED** [review]
The `implement` you referred to is without s. However, impelement in the sentence goes with "s". So is it a techincal term or not?

File: po/vi.po
Comment (vohoanglong0107) [reply po/vi.po:6217]: Haha, I agree with you that it's quite confused here. `implement` is being used here with a suffix of `s` because it was "following English grammar rule". However, I can assured you this is a technical term

Comment (vohoanglong0107) [reply po/vi.po:6217]: @qu-ngx If you are unsure, perharps this chapter in the Rust book (see [link](https://doc.rust-lang.org/book/ch10-02-traits.html#implementing-a-trait-on-a-type)) can help you make a decision.

Comment (vohoanglong0107) [reply po/vi.po:6217]: Ah, I should mention it, `implement` here is actually the expansion of `impl` keyword. That means `Vec implements Deref<Target = T>` is actually the same as `impl Deref<Target = T> for Vec`

Comment (qu-ngx) [reply po/vi.po:6217]: Yea, you better add it or put it somewhere in the brackets. And also if you don't translate the implements word then better keep the s with it as well so it should be implements instead of implement.

**Review by ghost — COMMENTED** [review]
I agree with your comment in the code review. When u are done changing, just ping me.

File: po/vi.po
Comment (vohoanglong0107) [reply po/vi.po:6268]: @qu-ngx Ah, I can see where the confusion coming from. Rust string has this rather strange property that, unlike other language like C++, when you access a string with `s[i]` syntax, the returned value could be an invalid character. The reason is, Rust's string is made of unicode characters, so each character could span multiple index. Suppose you have the string `s = "Xin chào"`. All the `X`, `i`, `n`, ` `, `c`, `h`, `o` are ASCII characters, therefore it need only one byte, consuming one slot in the string. However, `à` is a unicode character, so it need 2 bytes, consuming 2 slot. What this means is that when you try to access "Xin chào", Rust allow you to access `s[0..2]`, which prints out `Xi`, `s[5..8]`, which equal to `hà`, but you can't access `s[5..7]`, since 7 is in the middle of `à` character. This is what the excerpt refer to as the `character boundaries`.
However, I acknowledge that my translation poorly conveys that idea. I will see if I can improve this.

**Comment by vohoanglong0107** [comment]
@qu-ngx I think this should be ready for another go

**Review by tidefield — APPROVED** [review]
Sorry for the delay on reviewing the PRs. I've been on the road for the past few weeks.

**Review by ghost — APPROVED** [review]

**Pull request merged**
